### ISSUE TITLE
cmd/link: support relocations typical with -mcmodel=medium with internal linking

### ISSUE THREAD

**Issue by thanm** [golang/go#67475]
TLDR: it would useful if the Go linker could be changed to add support for relocations typically emitted by C compilers for code built with -mcmodel=medium.

More detail:

Consider this program (using txtar format):

```

-- go.mod --
module mcmodel.medium

go 1.23
-- main.go --
package main

import mcmodel "mcmodel.medium/mm"

func main() {
	println(mcmodel.Foo(1))
}
-- mm/bar.go --
package mcmodel

/*

#cgo CFLAGS: -mlarge-data-threshold=3  -mcmodel=medium

#include "defs.h"

LL ExtBigVar;

int ExtSmallVar;
*/
import "C"

func Bar(q int) int {
	return 43
}
-- mm/foo.go --
package mcmodel

/*

#cgo CFLAGS: -mlarge-data-threshold=3  -mcmodel=medium

#include "defs.h"

LL BigVar;

extern LL ExtBigVar;

int SmallVar;

extern int ExtSmallVar;

int ret1() {
  printf("%d %d\n", SmallVar, ExtSmallVar);
  return SmallVar + ExtSmallVar;
}

int ret2() {
  printf("%d %d\n", BigVar.x, ExtBigVar.x);
  return BigVar.x + ExtBigVar.x;
}
*/
import "C"

func Foo(x int) int {
	return Bar(x) + int(C.ret1()+C.ret2())
}
```

This code currently will not build due to "-mcmodel" etc not being on the allowlist for CGO cflags, but if you hack the go command to allow these options, then build:

```
# externl linking works
$ CC=cc go build .
# internal linking does not
$ CC=cc go build -ldflags=-linkmode=internal
# mcmodel.medium
$ WORK/b002/_pkg_.a(_x003.o): unknown relocation type 26; compiled without -fpic?
$
```

Note the "unknown relocation type 26". This is R_X86_64_GOTPC32; this reloc and R_X86_64_GOTOFF64 are frequently generated when building with the medium code model. 

It would be useful to ungrade the linker to handle these relocations; this is not a critical problem by any means since we can always fall back on external linking, but given the increased prevalence of "-mcmodel=medium" out in the wild for building very large programs, it might well be a good thing for us to add support for this.

NB: patch required to get to this to build with go command is below.

```
$ git diff
diff --git a/src/cmd/go/internal/work/security.go b/src/cmd/go/internal/work/security.go
index 8e788b0425..1c9d466189 100644
--- a/src/cmd/go/internal/work/security.go
+++ b/src/cmd/go/internal/work/security.go
@@ -117,6 +117,8 @@ var validCompilerFlags = []*lazyregexp.Regexp{
        re(`-mthumb(-interwork)?`),
        re(`-mthreads`),
        re(`-mwindows`),
+       re(`-mcmodel=(.+)`),
+       re(`-mlarge-data-threshold=(.+)`),
        re(`-no-canonical-prefixes`),
        re(`--param=ssp-buffer-size=[0-9]*`),
        re(`-pedantic(-errors)?`),
$ 
```
